### ISSUE TITLE
High CPU memory usage when calling GradientTape's gradient() when using multiple threads/cores

### ISSUE THREAD

**Issue by jammm** [tensorflow/tensorflow#41718]
**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): Yes
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Linux Ubuntu 18.04.3
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device: None
- TensorFlow installed from (source or binary): Binary (pip install tf-nightly)
- TensorFlow version (use command below): tf-nightly 2.4.0.dev20200724
- Python version: 3.6.9
- Bazel version (if compiling from source): None
- GCC/Compiler version (if compiling from source): None
- CUDA/cuDNN version: None (CPU only)
- GPU model and memory: None (CPU only)

**Describe the current behavior**
The memory usage when using multiple threads is ~7GB on the example code provided below.

**Describe the expected behavior**
The memory usage should be around ~850-900MB, which is what you get when you only use 1 thread via the following:
```
tf.config.threading.set_inter_op_parallelism_threads(1)
tf.config.threading.set_intra_op_parallelism_threads(1)
```

**Standalone code to reproduce the issue**
https://colab.research.google.com/drive/1XEcTR433ePm4-OTi3i22PshKUk2aFFq-?usp=sharing
Note that the issue is not reproducible in colab as it only runs on 1 core.

**Other info / logs** Include any logs or source code that would be helpful to
diagnose the problem. If including tracebacks, please include the full
traceback. Large logs and files should be attached.

Memory usage when tf.config.threading.set_*_op_parallelism_threads to 1:
```
2020-07-25 11:22:37.430254: I tensorflow/stream_executor/platform/default/dso_loader.cc:48] Successfully opened dynamic library libcudart.so.10.1
2020-07-25 11:23:01.051408: I tensorflow/stream_executor/platform/default/dso_loader.cc:48] Successfully opened dynamic library libcuda.so.1
2020-07-25 11:23:01.051508: E tensorflow/stream_executor/cuda/cuda_driver.cc:314] failed call to cuInit: UNKNOWN ERROR (-1)
2020-07-25 11:23:01.051596: I tensorflow/stream_executor/cuda/cuda_diagnostics.cc:156] kernel driver does not appear to be running on this host (big000): /proc/driver/nvidia/version does not exist
2020-07-25 11:23:01.052370: I tensorflow/core/platform/cpu_feature_guard.cc:142] This TensorFlow binary is optimized with oneAPI Deep Neural Network Library (oneDNN) to use the following CPU instructions in performance-critical operations:  AVX2 AVX512F FMA
To enable them in other operations, rebuild TensorFlow with the appropriate compiler flags.
2020-07-25 11:23:01.069108: I tensorflow/core/platform/profile_utils/cpu_utils.cc:104] CPU Frequency: 2094995000 Hz
2020-07-25 11:23:01.069468: I tensorflow/compiler/xla/service/service.cc:168] XLA service 0x5732470 initialized for platform Host (this does not guarantee that XLA will be used). Devices:
2020-07-25 11:23:01.069529: I tensorflow/compiler/xla/service/service.cc:176]   StreamExecutor device (0): Host, Default Version
2020-07-25 11:23:01.638778: I tensorflow/compiler/mlir/mlir_graph_optimization_pass.cc:118] None of the MLIR optimization passes are enabled (registered 1)
-------------------------------------------------------------------- -1.08216429
Memory usage:  867950592
-------------------------------------------------------------------- -1.02312386
Memory usage:  898691072
-------------------------------------------------------------------- -1.02274418
Memory usage:  899932160
-------------------------------------------------------------------- -0.958470643
Memory usage:  900022272
-------------------------------------------------------------------- -0.994939446
Memory usage:  900042752
-------------------------------------------------------------------- -1.04647171
Memory usage:  900050944
-------------------------------------------------------------------- -1.00102437
Memory usage:  900235264
-------------------------------------------------------------------- -0.987627268
Memory usage:  900382720
-------------------------------------------------------------------- -1.00234675
Memory usage:  900382720
-------------------------------------------------------------------- -0.957870305
Memory usage:  900390912
-------------------------------------------------------------------- -1.04384947
Memory usage:  900399104
-------------------------------------------------------------------- -1.05602765
Memory usage:  900399104
-------------------------------------------------------------------- -1.03467119
Memory usage:  900411392
-------------------------------------------------------------------- -1.00377405
Memory usage:  900415488
-------------------------------------------------------------------- -0.925668299
Memory usage:  900415488
-------------------------------------------------------------------- -1.08483529
Memory usage:  900423680
-------------------------------------------------------------------- -1.00288522
Memory usage:  900423680
-------------------------------------------------------------------- -0.909320414
Memory usage:  900427776
-------------------------------------------------------------------- -1.03661454
Memory usage:  900431872
-------------------------------------------------------------------- -0.955467224
...
Memory usage:  900431872
```

Memory usage peaks to around ~850MB.

Memory usage when *not* setting tf.config.threading.set_*_op_parallelism_threads to 1 i.e., when multiple threads/cores are being used:

```
2020-07-25 11:06:52.523261: I tensorflow/stream_executor/platform/default/dso_loader.cc:48] Successfully opened dynamic library libcudart.so.10.1
2020-07-25 11:07:17.286000: I tensorflow/stream_executor/platform/default/dso_loader.cc:48] Successfully opened dynamic library libcuda.so.1
2020-07-25 11:07:17.286099: E tensorflow/stream_executor/cuda/cuda_driver.cc:314] failed call to cuInit: UNKNOWN ERROR (-1)
2020-07-25 11:07:17.286161: I tensorflow/stream_executor/cuda/cuda_diagnostics.cc:156] kernel driver does not appear to be running on this host (big000): /proc/driver/nvidia/version does not exist
2020-07-25 11:07:17.287431: I tensorflow/core/platform/cpu_feature_guard.cc:142] This TensorFlow binary is optimized with oneAPI Deep Neural Network Library (oneDNN) to use the following CPU instructions in performance-critical operations:  AVX2 AVX512F FMA
To enable them in other operations, rebuild TensorFlow with the appropriate compiler flags.
2020-07-25 11:07:17.325674: I tensorflow/core/platform/profile_utils/cpu_utils.cc:104] CPU Frequency: 2094995000 Hz
2020-07-25 11:07:17.344702: I tensorflow/compiler/xla/service/service.cc:168] XLA service 0x504e590 initialized for platform Host (this does not guarantee that XLA will be used). Devices:
2020-07-25 11:07:17.344800: I tensorflow/compiler/xla/service/service.cc:176]   StreamExecutor device (0): Host, Default Version
2020-07-25 11:07:18.198284: I tensorflow/compiler/mlir/mlir_graph_optimization_pass.cc:118] None of the MLIR optimization passes are enabled (registered 1)
-------------------------------------------------------------------- -1.08216429
Memory usage:  933933056
-------------------------------------------------------------------- -1.02312386
Memory usage:  1463521280
-------------------------------------------------------------------- -1.02274418
Memory usage:  1981599744
-------------------------------------------------------------------- -0.958470643
Memory usage:  2500599808
-------------------------------------------------------------------- -0.994939446
Memory usage:  3082350592
-------------------------------------------------------------------- -1.04647171
Memory usage:  3530485760
-------------------------------------------------------------------- -1.00102437
Memory usage:  4058849280
-------------------------------------------------------------------- -0.987627268
Memory usage:  4124266496
-------------------------------------------------------------------- -1.00234675
Memory usage:  4574625792
-------------------------------------------------------------------- -0.957870305
Memory usage:  5169975296
-------------------------------------------------------------------- -1.04384947
Memory usage:  5198962688
-------------------------------------------------------------------- -1.05602765
Memory usage:  5294829568
-------------------------------------------------------------------- -1.03467119
Memory usage:  5808898048
-------------------------------------------------------------------- -1.00377405
Memory usage:  6403420160
-------------------------------------------------------------------- -0.925668299
Memory usage:  6428565504
-------------------------------------------------------------------- -1.08483529
Memory usage:  6433001472
-------------------------------------------------------------------- -1.00288522
Memory usage:  6433193984
-------------------------------------------------------------------- -0.909320414
Memory usage:  6462169088
-------------------------------------------------------------------- -1.03661454
Memory usage:  6982217728
-------------------------------------------------------------------- -0.955467224
Memory usage:  6985584640
-------------------------------------------------------------------- -1.02524447
Memory usage:  7088070656
-------------------------------------------------------------------- -1.06649399
Memory usage:  7114588160
-------------------------------------------------------------------- -0.991624892
Memory usage:  7116353536
-------------------------------------------------------------------- -1.01972055
Memory usage:  7143043072
-------------------------------------------------------------------- -0.994572461
Memory usage:  7144923136
-------------------------------------------------------------------- -0.965809107
Memory usage:  7149711360
-------------------------------------------------------------------- -1.02491224
Memory usage:  7181701120
-------------------------------------------------------------------- -1.02480865
Memory usage:  7723294720
-------------------------------------------------------------------- -0.970712662
Memory usage:  7724584960
-------------------------------------------------------------------- -1.02038872
Memory usage:  7726415872
-------------------------------------------------------------------- -1.01687396
Memory usage:  7726661632
-------------------------------------------------------------------- -1.03144
Memory usage:  7726739456
-------------------------------------------------------------------- -0.950338304
Memory usage:  7748882432
-------------------------------------------------------------------- -0.9357301
Memory usage:  7748997120
-------------------------------------------------------------------- -0.932868242
Memory usage:  7781220352
-------------------------------------------------------------------- -0.950198233
Memory usage:  7781298176
-------------------------------------------------------------------- -0.966640532
Memory usage:  7782912000
-------------------------------------------------------------------- -0.97049737
Memory usage:  7783198720
-------------------------------------------------------------------- -1.00543392
Memory usage:  7810494464
-------------------------------------------------------------------- -0.949840665
Memory usage:  7810592768
-------------------------------------------------------------------- -1.0022217
Memory usage:  7810596864
-------------------------------------------------------------------- -1.12232506
Memory usage:  7810686976
-------------------------------------------------------------------- -1.01822805
Memory usage:  7813079040
-------------------------------------------------------------------- -1.04360735
Memory usage:  7813513216
-------------------------------------------------------------------- -0.956581
Memory usage:  7813517312
-------------------------------------------------------------------- -0.981084287
Memory usage:  7813537792
-------------------------------------------------------------------- -0.977599561
Memory usage:  7813615616
-------------------------------------------------------------------- -1.01713431
Memory usage:  8277536768
-------------------------------------------------------------------- -1.03842521
Memory usage:  8276979712
-------------------------------------------------------------------- -0.971421421
Memory usage:  8299802624
-------------------------------------------------------------------- -1.06472087
Memory usage:  8299806720
-------------------------------------------------------------------- -1.00037909
Memory usage:  8299974656
-------------------------------------------------------------------- -1.00314832
Memory usage:  8299995136
-------------------------------------------------------------------- -0.99641335
Memory usage:  8299999232
-------------------------------------------------------------------- -1.07826197
Memory usage:  8300007424
-------------------------------------------------------------------- -0.998241484
Memory usage:  8300498944
-------------------------------------------------------------------- -1.01393604
Memory usage:  8300503040
-------------------------------------------------------------------- -1.04430008
Memory usage:  8301498368
-------------------------------------------------------------------- -0.956604362
Memory usage:  8303042560
-------------------------------------------------------------------- -0.994233251
Memory usage:  8303177728
-------------------------------------------------------------------- -1.0575521
Memory usage:  8303247360
-------------------------------------------------------------------- -0.975697637
Memory usage:  8304357376
-------------------------------------------------------------------- -1.0441494
Memory usage:  8304455680
-------------------------------------------------------------------- -0.949515164
Memory usage:  8304496640
-------------------------------------------------------------------- -0.935756862
Memory usage:  8304496640
-------------------------------------------------------------------- -1.00150812
Memory usage:  8304496640
-------------------------------------------------------------------- -0.983478725
Memory usage:  8304627712
-------------------------------------------------------------------- -0.986330569
Memory usage:  8304635904
-------------------------------------------------------------------- -1.13513744
Memory usage:  8287686656
-------------------------------------------------------------------- -1.03896141
Memory usage:  8287719424
-------------------------------------------------------------------- -0.988004625
Memory usage:  8287723520
-------------------------------------------------------------------- -0.985565066
Memory usage:  8287731712
-------------------------------------------------------------------- -1.0104388
Memory usage:  8304316416
-------------------------------------------------------------------- -1.06218076
Memory usage:  8304353280
-------------------------------------------------------------------- -0.94707495
Memory usage:  8304357376
-------------------------------------------------------------------- -0.998919964
Memory usage:  8304357376
-------------------------------------------------------------------- -0.974167764
Memory usage:  8304361472
-------------------------------------------------------------------- -0.97153908
Memory usage:  8304365568
-------------------------------------------------------------------- -0.964009821
Memory usage:  8304369664
-------------------------------------------------------------------- -0.992211759
Memory usage:  8304373760
-------------------------------------------------------------------- -1.04404831
Memory usage:  8304431104
-------------------------------------------------------------------- -1.02076721
Memory usage:  8304435200
-------------------------------------------------------------------- -0.962560713
Memory usage:  8304439296
-------------------------------------------------------------------- -1.02682006
Memory usage:  8304443392
-------------------------------------------------------------------- -1.03967941
Memory usage:  8304447488
-------------------------------------------------------------------- -1.07996821
Memory usage:  8304480256
-------------------------------------------------------------------- -0.98360759
Memory usage:  8305147904
-------------------------------------------------------------------- -1.0019182
Memory usage:  8305152000
-------------------------------------------------------------------- -0.935099125
Memory usage:  8305205248
-------------------------------------------------------------------- -1.03370762
Memory usage:  8305209344
-------------------------------------------------------------------- -0.98182
Memory usage:  8305213440
-------------------------------------------------------------------- -0.964839816
Memory usage:  8305217536
-------------------------------------------------------------------- -0.926978409
Memory usage:  8305221632
-------------------------------------------------------------------- -0.972101748
...
```

As you can see, memory usage peaks to around ~8GB.

The issue cannot be reproduced on google colab because colab only runs on 1 CPU anyway, so the behavior is the same regardless of whether you call tf.config.threading.set_*_op_parallelism_threads(1) or not.

**Comment by jammm** [comment]
I have a feeling that this issue is related to the `model.fit()` high memory usage as well.
https://github.com/tensorflow/tensorflow/issues/40523

**Comment by ravikyram** [comment]
I have tried in colab with TF nightly version(`2.4.0-dev20200726`).Please, find the gist [here](https://colab.research.google.com/gist/ravikyram/e42914b14ec8d98204c47178215331c9/untitled183.ipynb).Thanks!

**Comment by jammm** [comment]
I could try debugging this myself, but I'm not sure where to start looking. If any TF devs are reading this, it would be great if you could point me to the code that could be causing this issue.

For what it's worth, the issue disappears when you comment out the "tape.gradient" line.

**Comment by tomerk** [comment]
Quick triage notes:
* This doesn't use keras (even though the repro includes a keras backend clear_session call)
* The repro example appears to use: tf.function, gradienttape, autograph, tensorarrays, iteration. The memory usage appears to be a problem when using multiple threads.

@jammm , some quick questions for you:
1. are you observing that the memory usage is proportional to the number of threads? Or does it skyrocket to 8 GB as soon as you increase parallelism at all?
2. if you explicitly call `gc.collect()`, what happens to the reported memory? (so we know whether the issue is an actual leak or a reference cycle that python's gc can eventually handle)
3. Can you reproduce your issue when taking out any of the factors mentioned above? (e.g. tensorarrays, the tf.function, the loop, etc.) or does it only happen when all of them are combined?

**Comment by rjdbcm** [comment]
Was able to fix this by adding the following within my `with GradientTape() as t:` context:
`
variables = self.trainable_variables
gradients = t.gradient(loss, variables)
self.optimizer.apply_gradients(zip(gradients, variables))
del gradients
del variables
gc.collect()
`

**Comment by rmothukuru** [comment]
@jammm,
Can you please respond to  [tomerk's comment](https://github.com/tensorflow/tensorflow/issues/41718#issuecomment-700957911) and [rjdbcm's comment](https://github.com/tensorflow/tensorflow/issues/41718#issuecomment-731600429)? Thanks!

**Comment by jammm** [comment]
@rmothukuru I haven't had the chance to get back to this yet. I remember decreasing the scale of the training itself so it would keep leaking memory but at a slow enough rate to finish before using too much RAM.

**Comment by Venkat6871** [comment]
Hi,

Thank you for opening this issue. Since this issue has been open for a long time, the code/debug information for this issue may not be relevant with the current state of the code base.

The Tensorflow team is constantly improving the framework by fixing bugs and adding new features. We suggest you try the latest TensorFlow [version](https://pypi.org/project/tensorflow/) with the latest compatible hardware configuration which could potentially resolve the issue. If you are still facing the issue, please create a new GitHub issue with your latest findings, with all the debugging information which could help us investigate.

Please follow the [release notes](https://github.com/tensorflow/tensorflow/blob/master/RELEASE.md) to stay up to date with the latest developments which are happening in the Tensorflow space.
